### PR TITLE
Issue #882: update LogicConditionNeedOptimizationCheck to avoid false positive

### PR DESCRIPTION
**Pull request by nrmancuso**
From https://github.com/checkstyle/checkstyle/pull/11100

Discovered in report at https://nmancus1.github.io/issue-10848/issue-10848_check_diff_reports_2021_12_31/diff_sevntu-check-regression_part_2/index.html

Builds will fail until 9.3 release.

### CODE REVIEW

**Comment by rnveach** [comment]
> Builds will fail until 9.3 release.

**Comment by rnveach** [comment]
@nrmancuso @romani Is this PR still blocked?
https://github.com/checkstyle/eclipse-cs/issues/339 is merged.

**Comment by romani** [comment]
rebased on latest, but there are problems in UT execution

**Comment by rnveach** [comment]
````
[ERROR] Tests run: 8, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 3.701 s <<< FAILURE! - in com.github.sevntu.checkstyle.internal.AllChecksTest

[ERROR] testAllInputsHaveTest  Time elapsed: 0.148 s  <<< FAILURE!
org.opentest4j.AssertionFailedError: Resource must be named after a Test like 'InputMyCheck.java' and be in the same package as the test: com/github/sevntu/checkstyle/checks/coding/InputLogicConditionNeedsOptimizationPatterns.java ==> expected: <true> but was: <false>
	at com.github.sevntu.checkstyle.internal.AllChecksTest.verifyInputFile(AllChecksTest.java:290)
	at com.github.sevntu.checkstyle.internal.AllChecksTest.lambda$testAllInputsHaveTest$4(AllChecksTest.java:185)
	at com.github.sevntu.checkstyle.internal.AllChecksTest.testAllInputsHaveTest(AllChecksTest.java:184)
````

**Comment by rnveach** [comment]
````
[INFO] --- jacoco-maven-plugin:0.8.7:check (default-check) @ sevntu-checks ---
[WARNING] Rule violated for class com.github.sevntu.checkstyle.checks.coding.LogicConditionNeedOptimizationCheck: branches covered ratio is 0.97, but expected minimum is 1.00
````

File: sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/LogicConditionNeedOptimizationCheck.java [hunk 1]
```diff
@@ -190,4 +190,18 @@ private static boolean branchContains(DetailAST start, DetailAST end, int type)
...
+     *  If pattern variable is introduced, 'instanceof` must appear before condition
+     *  that checks pattern variable.
+     *
+     * @param logicNode logic node whose children we check
+     * @return true if a pattern variable is introduced
+     */
+    private static boolean isPatternVariableIntroduced(DetailAST logicNode) {
+        final DetailAST firstOperand = logicNode.getFirstChild();
+        return firstOperand.getType() == TokenTypes.LITERAL_INSTANCEOF
+                        && firstOperand.getLastChild() != null
```
Comment (rnveach) [inline sevntu-checks/src/main/java/com/github/sevntu/checkstyle/checks/coding/LogicConditionNeedOptimizationCheck.java:203]: This 1 line is missing branch coverage.

**Review by rnveach — CHANGES_REQUESTED** [review]

**Comment by rnveach** [comment]
@nrmancuso Can we finish this?

**Comment by nrmancuso** [comment]
> @nrmancuso Can we finish this?

We should close this as syntax for pattern matching in switch has changed. We (I) made a bad decision to support a preview feature, and we will need to update grammar in main repo as this syntax is not valid any longer. I will leave this open until I create an issue in the main repo to explain all.

**Comment by nrmancuso** [comment]
I have opened an issue for syntax update at https://github.com/checkstyle/checkstyle/issues/12515, we know that changes similar to what is in this PR will be required in that update.

**Pull request closed without merging**